File: backend/app/model/resource.py
from sqlalchemy import func, Column, Integer, String, DateTime, Boolean, ForeignKey
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.orm import relationship

from app.database import Base
from app.model.category import Category


class Resource(Base):
    __tablename__ = "resource"
    __label__ = "Online Information"
    id = Column(Integer, primary_key=True)
    type = Column(String)
    title = Column(String)
    last_updated = Column(DateTime(timezone=True), default=func.now())
    description = Column(String)
    insurance = Column(String)
    organization_name = Column(String)
    phone = Column(String)
    phone_extension = Column(String)
    website = Column(String)
    contact_email = Column(String)
    video_code = Column(String)
    is_uva_education_content = Column(Boolean)
    is_draft = Column(Boolean)
    ages = Column(ARRAY(String), default=[])
    languages = Column(ARRAY(String), default=[])
    covid19_categories = Column(ARRAY(String), default=[])
    should_hide_related_resources = Column(Boolean, default=False)

    __mapper_args__ = {"polymorphic_identity": "resource", "polymorphic_on": type}

    def indexable_content(self):
        return " ".join(
            filter(
                None,
                (
                    self.title,
                    self.description,
                    self.insurance,
                    self.category_names(),
                ),
            )
        )

    def category_names(self):
        cat_text = ""
        for cat in self.categories:
            cat_text = cat_text + " " + cat.category.name

        if self.ages is not None and len(self.ages) > 0:
            cat_text = cat_text + " " + " ".join(self.ages)
        if self.languages is not None and len(self.languages) > 0:
            cat_text = cat_text + " " + " ".join(self.languages)
        if self.covid19_categories is not None and len(self.covid19_categories) > 0:
            cat_text = cat_text + " " + " ".join(self.covid19_categories)

        return cat_text


class ResourceCategory(Base):
    __tablename__ = "resource_category"
    id = Column(Integer, primary_key=True)
    last_updated = Column(DateTime(timezone=True), default=func.now())
    type = Column(String)
    resource_id = Column(Integer, ForeignKey(Resource.id), nullable=False)
    category_id = Column(Integer, ForeignKey(Category.id), nullable=False)
    resource = relationship(Resource, backref="resource_categories")
    category = relationship(Category, backref="category_resources")


Resource.categories = relationship("ResourceCategory", back_populates="resource")

File: frontend/src/app/_models/resource.ts
import {Category} from './category';
import {GeoLocation} from './geolocation';
import {ResourceCategory} from './resource_category';

export enum ResourceType {
  RESOURCE = 'resource',
  EVENT = 'event',
  LOCATION = 'location',
}

export interface ResourceProps {
  type: ResourceType;
  title: string;
  description: string;
  post_event_description: string;
  insurance: string;
  date?: string;
  time?: string;
  ticket_cost?: string;
  organization_name?: string;
  primary_contact?: string;
  location_name?: string;
  street_address1?: string;
  street_address2?: string;
  city?: string;
  state?: string;
  zip?: string;
  phone?: string;
  phone_extension?: string;
  website?: string;
  contact_email?: string;
  video_code?: string;
  is_uva_education_content?: boolean;
  is_draft?: boolean;
  last_updated?: string;
  status?: string;
  resource_categories?: ResourceCategory[];
  categories?: Category[];
  ages?: string[];
  languages?: string[];
  covid19_categories?: string[];
  includes_registration?: boolean;
  webinar_link?: string;
  post_survey_link?: string;
  max_users?: number;
  image_url?: string;
  registration_url?: string;
  should_hide_related_resources?: boolean;
}

export class Resource extends GeoLocation {
  type: ResourceType;
  title: string;
  description: string;
  post_event_description: string;
  insurance: string;
  date?: string;
  time?: string;
  ticket_cost?: string;
  organization_name?: string;
  primary_contact?: string;
  location_name?: string;
  street_address1?: string;
  street_address2?: string;
  city?: string;
  state?: string;
  zip?: string;
  phone?: string;
  phone_extension?: string;
  website?: string;
  contact_email?: string;
  video_code?: string;
  is_uva_education_content?: boolean;
  is_draft?: boolean;
  last_updated?: string;
  status?: string;
  resource_categories?: ResourceCategory[];
  categories?: Category[];
  ages?: string[];
  languages?: string[];
  covid19_categories?: string[];
  includes_registration?: boolean;
  webinar_link?: string;
  post_survey_link?: string;
  max_users?: number;
  image_url?: string;
  registration_url?: string;
  should_hide_related_resources?: boolean;

  constructor(private _props: Partial<ResourceProps>) {
    super(_props);

    for (const propName in this._props) {
      if (this._props.hasOwnProperty(propName)) {
        this[propName] = this._props[propName];
      }
    }
  }
}
